controllers/serviceclass_controller.go

```diff
@@ -47,16 +71,115 @@ type ServiceClassReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
 func (r *ServiceClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	reconcileLog := log.FromContext(ctx)
+
+	serviceClass := v1alpha1.ServiceClass{}
+	err := r.Get(ctx, req.NamespacedName, &serviceClass)
+	if err != nil {
+		reconcileLog.Error(err, "Failed to retrieve ServiceClass", "namespace", req.Namespace, "name", req.Name)
+		return ctrl.Result{}, nil
+	}
+
+	typemeta := metav1.TypeMeta{
+		Kind:       serviceClass.Spec.Resource.Kind,
+		APIVersion: serviceClass.Spec.Resource.APIVersion,
+	}
+	gvk := typemeta.GroupVersionKind()
+	mapping, err := r.Mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		reconcileLog.Error(err, "Failed to retrieve resource type", "gvk", gvk)
+		return ctrl.Result{}, nil
+	}
+
+	services, err := r.Interface.Resource(mapping.Resource).
+		Namespace(serviceClass.Namespace).
+		List(ctx, metav1.ListOptions{})
+
+	if err != nil || services == nil {
+		reconcileLog.Error(err, "Failed to retrieve resources", "gvr", mapping.Resource)
+		return ctrl.Result{}, nil
+	}
+
+	err = r.CreateRegisteredServices(ctx, serviceClass, *services)
+	if err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	return ctrl.Result{}, err
+}
+
+func (r *ServiceClassReconciler) CreateRegisteredServices(ctx context.Context, serviceClass v1alpha1.ServiceClass, services unstructured.UnstructuredList) error {
+	l := log.FromContext(ctx)
+	mappings := map[string]*jsonpath.JSONPath{}
+	for _, mapping := range serviceClass.Spec.Resource.ServiceEndpointDefinitionMapping {
+		path := jsonpath.New("")
+		err := path.Parse(fmt.Sprintf("{%s}", mapping.JsonPath))
+		if err != nil {
+			return err
+		}
+		mappings[mapping.Name] = path
+	}
+
+	for _, data := range services.Items {
+		sedMappings, err := LookupServiceEndpointDescriptor(mappings, data)
+		if err != nil {
+			l.Error(err, "Failed to lookup service endpoint descriptor values", "name", data.GetName(), "namespace", data.GetNamespace(), "gvk", data.GroupVersionKind())
```

Comment:
we can give the user some insights on this error by adding a condition on SED generation to the RegisteredService's status. what do you think?

Reply:
I think I'll punt on this for a future PR; I want to do some design on what's useful to expose as a status condition.